### PR TITLE
Add YAML highlighter

### PR DESCRIPTION
**Pull request by hason**
Replaces #408

### CODE REVIEW

File: examples/prism-yaml.html [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+<h2>Null and Boolean</h2>
+<pre><code>---
+A null: null
+A null: ~
+Also a null: # Empty
+Not a null: ""
+Booleans: [ true, True, false, FALSE ]
+</code></pre>
+
+<h2>Numbers a timestamps</h2>
```
Comment (Golmote) [inline examples/prism-yaml.html]: You meant "Numbers and timestamps", right?

File: components/prism-yaml.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+Prism.languages.yaml = {
+	'scalar': {
+		pattern: /([\-:]\s*(![^\s]+)?[ \t]*[\|>])[ \t]*(?:(\n[ \t]+)[^\r\n]+)?(?:\3[^\r\n]+)*/,
```
Comment (Golmote) [inline components/prism-yaml.js]: If I read this correctly, the `\3` refers to the group `(\n[ \t]+)` which is part of an optional group. Shoudn't the last bit `(?:\3[^\r\n]+)*` be included in the same optional group (as the `\3` is mandatory)?

File: components/prism-yaml.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		lookbehind: true,
+		alias: 'string'
+	},
+	'comment': /#[^\n]+/,
+	'key': {
+		pattern: /(\s*[:\-,\[\{\n?][ \t]*(![^\s]+)?[ \t]*)[^\n\{\[\]\},#]+?(?=\s*:\s)/i,
```
Comment (Golmote) [inline components/prism-yaml.js]: There is no need to escape the braces `{` and `}` inside character classes.

Comment (Golmote) [inline components/prism-yaml.js]: Also, is the `i` flag actually useful?

File: components/prism-yaml.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		lookbehind: true,
+		alias: 'atrule'
+	},
+	'directive': {
+		pattern: /(^|\n)[ \t]*%[^\n]+/,
+		lookbehind: true,
+		alias: 'important'
+	},
+	'datetime': {
+		pattern: /([:\-,\[\{]\s*(![^\s]+)?[ \t]*)(\d{4}-\d\d?-\d\d?(t|[ \t]+)\d\d?:\d{2}:\d{2}(\.\d*)?[ \t]*(Z|[-+]\d\d?(:\d{2})?)?|\d{4}-\d{2}-\d{2}|\d\d?:\d{2}(:\d{2}(\.\d*)?)?)(?=[ \t]*(\n|$|,|\]|\}))/i,
```
Comment (Golmote) [inline components/prism-yaml.js]: Here again, no need to escape the curly brace in the first character class.

File: components/prism-yaml.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	},
+	'string': {
+		pattern: /([:\-,\[\{]\s*(![^\s]+)?[ \t]*)("(\\\\|\\"|[^"\\])*"|'(\\\\|\\'|[^'\\])*')(?=[ \t]*(\n|$|,|\]|\}))/,
+		lookbehind: true
+	},
+	'number': {
+		pattern: /([:\-,\[\{]\s*(![^\s]+)?[ \t]*)[\+\-]?(0x[\dA-Fa-f]+|0o[0-7]+|(\d+\.?\d*|\.?\d+)(e[\+\-]?\d+)?|\.inf|\.nan)[ \t]*(?=\n|$|,|\]|\})/i,
+		lookbehind: true
+	},
+	'tag': /![^\s]+/i,
```
Comment (Golmote) [inline components/prism-yaml.js]: Is the `i` flag useful here?

File: components/prism-yaml.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	'important': /[&\*][\w]+/,
```
Comment (Golmote) [inline components/prism-yaml.js]: The `*` does not need to be escaped inside a character class.

File: components/prism-yaml.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	'punctuation': /([:\[\]\{\}\-,\|>?]|---|\.\.\.)/
```
Comment (Golmote) [inline components/prism-yaml.js]: The curly braces `{}` and the pipe `|` do not need to be escaped inside a character class.

**Comment by hason** [comment]
@Golmote @LeaVerou What do you think?

File: components/prism-yaml.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+Prism.languages.yaml = {
+	'scalar': {
+		pattern: /([\-:]\s*(![^\s]+)?[ \t]*[\|>])[ \t]*(?:(\n[ \t]+)[^\r\n]+(?:\3[^\r\n]+)*)/,
```
Comment (Golmote) [inline components/prism-yaml.js]: I missed one here: the `|` does not need to be escaped inside character classes.

**Comment by Golmote** [comment]
It looks good to me. Now it's up to @LeaVerou to decide.

**Comment by LeaVerou** [comment]
LGTM, merged, thanks! :)

**Pull request merged**